### DISCUSSION TITLE
Declarative Dataclass Mapping with runtime type checks

### DISCUSSION THREAD

**Discussion by BayerSe** [sqlalchemy/sqlalchemy#9264] [Usage Questions]
Hi,

we're currently using the now deprecated [Mapping dataclasses using Declarative Mapping](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#mapping-dataclasses-using-declarative-mapping) with [pydantic dataclasses](https://docs.pydantic.dev/usage/dataclasses/).

This allows us e.g. the following:

```python
from dataclasses import field
from pydantic.dataclasses import dataclass

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy.orm import registry

mapper_registry = registry()


@mapper_registry.mapped
@dataclass
class User:
    __tablename__ = "user"

    __sa_dataclass_metadata_key__ = "sa"
    id: int = field(metadata={"sa": Column(Integer, primary_key=True)})

User(id="A")
```
which raises 

```python
pydantic.error_wrappers.ValidationError: 1 validation error for User
id
  value is not a valid integer (type=type_error.integer)
```
which is not the case when using the standard library dataclasses instead of pydantic's version.

Given that the above approach is now deprecated, I would like to make use of the [novel way](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#orm-declarative-native-dataclasses):

```python
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column


class Base(MappedAsDataclass, DeclarativeBase):
    """subclasses will be converted to dataclasses"""


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
```    

Is there any way to combine the new way with pydantic's runtime type checking, so that `User(id="A")`  raises?

**Comment by zzzeek** [comment]
hey there -

I didnt know pydantic's dataclasses feature was usable with SQLAlchemy at all, last I checked it was not compatible.  For support of our new feature, we use regular Python dataclasses, im not sure if pydantic has their own notion of this.    this would all depend on how pydantic provides dataclasses functionality so you'd need to start with them.

**Reply by zzzeek** [reply]
like if pydantic has `pydantic.dataclass`, that would need to be injected into SQLAlchemy's declarative.   But if that class decorator behaves wildly different from the builtin dataclass decorator, there would be a lot of problems that someone would need to work out, we could at most provide a hook for an alternate "dataclass" function but we can't develop/test pydantic-specific use cases on this end.

**Reply by BayerSe** [reply]
Thank you for responding so quickly, @zzzeek 

Yes, the compatibility of pydantic and sqlalchemy was not working at some point but it now works, see https://github.com/pydantic/pydantic/issues/2924

Some hook would be fantastic, especially since pydantic's dataclass aims to be a drop in replacement.

**Reply by zzzeek** [reply]
I'd be semi-surprised if a simple dropin "just works".    

Go right into the source and try it out, this is where we'd put a hook:


```diff
diff --git a/lib/sqlalchemy/orm/decl_base.py b/lib/sqlalchemy/orm/decl_base.py
index aeed9b4395..7f21109a52 100644
--- a/lib/sqlalchemy/orm/decl_base.py
+++ b/lib/sqlalchemy/orm/decl_base.py
@@ -1111,7 +1111,11 @@ class _ClassScanMapperConfig(_MapperConfig):
     ) -> None:
         cls._assert_dc_arguments(dataclass_setup_arguments)
 
-        dataclasses.dataclass(
+        special_dataclass_callable = getattr(
+            klass, "special_dataclass_callable", dataclasses.dataclass
+        )
+
+        special_dataclass_callable(
             klass,
             **{
                 k: v
```

then make a class like this:

```py
import pydantic

class MyClass(MappedAsDataclass):
    special_dataclass_callable = pydantic.dataclass
```

let me know how it goes.   We can add a similar hook as the one above but we usually will not be able to support major incompatibilities in pydantic if there are any

**Reply by BayerSe** [reply]
Thank you for the pointer. The simple example actually works perfectly with pydantic 1.10.4!

```python
import pydantic
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column


class MappedAsPydanticDataclass(MappedAsDataclass):
    special_dataclass_callable = pydantic.dataclasses.dataclass


class Base(MappedAsPydanticDataclass, DeclarativeBase):
    """subclasses will be converted to dataclasses"""


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)


user = User(id="A")
```

raises 

```python
pydantic.error_wrappers.ValidationError: 1 validation error for User
id
  value is not a valid integer (type=type_error.integer)
```


When I adjust the source per your suggestion.

**Reply by zzzeek** [reply]
OK....and....maybe spend some time with it, make sure nothing surprising happens w/ persistence / querying / etc ?

**Reply by zzzeek** [reply]
@Caselit any thoughts on the best way to hook like this?  maybe use an `@event` hook ?  or ``__dataclasses_callable__`` attribute?

**Reply by BayerSe** [reply]
Thank you for considering including it in sqlalchemy, @zzzeek ! I'll update the code base with the change and report back tomorrow.

**Reply by CaselIT** [reply]
> @CaselIT any thoughts on the best way to hook like this? maybe use an `@event` hook ? or `__dataclasses_callable__` attribute?

I would go with the attribute, since it's more similar to how the rest of the customization in the declarative base is done,

The issue with using an attribute in this case is that python will convert all callables to method when instancing the method, but maybe we could instruct people to use `staticmethod(...)` with it?

Maybe a better alternative could be to make it an attribute of the registry, so that it would not create this issue (and people could use it also when using the `mapped_as_dataclass` decorator). So with declarative you would do:
```py
class Base(DeclarativeBase):
  registry = registry(dataclasses_callable=my_dataclass_thing)
```

**Reply by zzzeek** [reply]
I think registry is the best approach here.   the one awkwardness is that the `MappedAsDataclass`  / `@as_dataclass` constructs can be used on a per-class or per-hierarchy basis, so as you say, maybe it could be part of the decorator also, though I think `MappedAsDataclass` is more of the "mainstream" API.    I think add it to registry for now and then see if people really wanted to mix two different dataclass libraries in one hiearchy, I dont think it's urgent to support that.

**Reply by CaselIT** [reply]
yes, makes sense to me. do you want to open an issue to track this?

**Reply by CaselIT** [reply]
I've created https://github.com/sqlalchemy/sqlalchemy/issues/9266.

One question is if we need to pass the complete module, or only the `dataclass` callable? If only the callable, then we cannot use `field()` internally, (not sure if we do use it)

**Reply by CaselIT** [reply]
> One question is if we need to pass the complete module, or only the `dataclass` callable? If only the callable, then we cannot use `field()` internally, (not sure if we do use it)

well, the dataclasses module in pydantic does not have the same interface of the builtin module, in particular it has no `field`, so the answer is that just `dataclass` is passed

**Reply by BayerSe** [reply]
@zzzeek , I've now tested this extensively. In our code base, I was able to identify one problem with replacing the standard dataclass with pydantic's version: when adding columns to the base class.


This example works perfectly fine with standard dataclasses but not with pydantic:

```python
import pydantic
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    MappedAsDataclass,
    mapped_column,
)



class MappedasPydanticDataclass(MappedAsDataclass):
    special_dataclass_callable = pydantic.dataclasses.dataclass


class Base(MappedasPydanticDataclass, DeclarativeBase):
    name: Mapped[str]


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
```

pydantic raises:

```python
TypeError: Fields of type "<class 'sqlalchemy.orm.base.Mapped'>" are not supported.
```
Other than that, it really just worked. Once the hook is available, I'd approach pydantic about this.

Do you want me to do further tests or experiments?

**Reply by CaselIT** [reply]
interesting, I wonder if it's a bug in sqlalchemy or something strange with pydantic.

**Reply by BayerSe** [reply]
@CaselIT , The following example raises the same error as above

```python
from pydantic.dataclasses import dataclass
from sqlalchemy.orm import Mapped

@dataclass
class Test:
    mapped: Mapped[int]
```

and can be healed by allowing arbitrary types:

```python
@dataclass(config=dict(arbitrary_types_allowed=True))
class Test:
    mapped: Mapped[int]
```

However, I only wonder why pydantic does *not* raise when I add the column to the subclasses. I guess the "unpacking" of the Mapped type is done before calling the `dataclass` method? Is there probably some issue with "unpacking" columns in the base class? The Mapped-type should never reach the dataclass, right?


After some testing:
When I have a column in the base class, then this column (and only this column) is contained in klass.__dataclass_fields__ right before calling dataclasses.dataclass(...) in method _apply_dataclasses_to_any_class

**Reply by CaselIT** [reply]
yes, that seems like a bug in sqlalchemy, thanks for testing

**Reply by zzzeek** [reply]
if pydantic is scanning the actual attributes of the superclass, and not just looking at the existing "__dataclass_fields__", that's the only way I can see it suddenly seeing that `Mapped` attribute there.     Seems like pydantic bug to me but I haven't dug in.

**Reply by zzzeek** [reply]
we would also assume/ hope that all the standard dataclass methods like `dataclasses.is_dataclass()` and `dataclasses.fields()` work **exactly** the same way when using pydantic.    If not, then that would cause problems too.

**Reply by zzzeek** [reply]
HMMm OK I can see what we are doing differently here.     for declarative mixins we aren't applying Field().   Let me see what's involved there

**Reply by zzzeek** [reply]
So you're going to want to move over to #9266 and if you can navigate the gerrit review at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4439 and try it out , that would be super helpful.

**Reply by BayerSe** [reply]
Hi @zzzeek,

with your changes, this snippet raises, which does not happen in 2.0.3

```python
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column
from typing import get_type_hints

class Base(MappedAsDataclass, DeclarativeBase):
    ...

class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)

user = User(id=0)
assert get_type_hints(user)["id"] == int
```

This came up as I'm actually testing the correspondence of python and sql types, for which I'm using `get_type_hints`.



Apart from that, I could not find other issues.

**Reply by BayerSe** [reply]
Oh, one more thing. Running `mypy ` with the strict flag shows:


```
test.py:7: error: Unexpected keyword argument "dataclass_callable" for "__init_subclass__" of "object"  [call-arg]
test.py:8: error: Class cannot subclass "MappedAsDataclass" (has type "Any")  [misc]
test.py:9: error: Class cannot subclass "DeclarativeBase" (has type "Any")  [misc]
```

**Reply by CaselIT** [reply]
> Hi @zzzeek,
> 
> with your changes, this snippet raises, which does not happen in 2.0.3
> 
> ```python
> from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column
> from typing import get_type_hints
> 
> class Base(MappedAsDataclass, DeclarativeBase):
>     ...
> 
> class User(Base):
>     __tablename__ = "user"
> 
>     id: Mapped[int] = mapped_column(primary_key=True)
> 
> user = User(id=0)
> assert get_type_hints(user)["id"] == int
> ```
> 
> This came up as I'm actually testing the correspondence of python and sql types, for which I'm using `get_type_hints`.
> 
> Apart from that, I could not find other issues.

what error does it raise with the proposed changes?



> Oh, one more thing. Running `mypy ` with the strict flag shows:
> 
> ```
> test.py:7: error: Unexpected keyword argument "dataclass_callable" for "__init_subclass__" of "object"  [call-arg]
> test.py:8: error: Class cannot subclass "MappedAsDataclass" (has type "Any")  [misc]
> test.py:9: error: Class cannot subclass "DeclarativeBase" (has type "Any")  [misc]
> ```

can you ensure you are using mypy version 1.0?

**Reply by BayerSe** [reply]
Sorry, I forgot the output:

```python
> get_type_hints(user)["id"]
sqlalchemy.orm.base.Mapped[int]
```
and `get_type_hints` gets this from here:
```python
>  user.__annotations__
{'id': sqlalchemy.orm.base.Mapped[int]}
```

Regarding `mypy` - oh I see. There's something strange on my end, I use `mypy` via `pre-commit`. The warnings do not appear if I use `mypy` directly.

**Reply by CaselIT** [reply]
@zzzeek I'm not sure if that change is wanted. I think so, since we are explicitly restoring the annotation dict of the class, just confirming

> Regarding mypy - oh I see. There's something strange on my end, I use mypy via pre-commit. The warnings do not appear if I use mypy directly.

It may be using a different version. IIRC since 2.0.3 it requires v1 since it uses `Self` that is not understood by previous versions.

**Reply by zzzeek** [reply]
> @zzzeek I'm not sure if that change is wanted. I think so, since we are explicitly restoring the annotation dict of the class, just confirming

what change?  thread is too long.  




> Hi @zzzeek,
> 
> with your changes, this snippet raises, which does not happen in 2.0.3
> 
> ```python
> from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column
> from typing import get_type_hints
> 
> class Base(MappedAsDataclass, DeclarativeBase):
>     ...
> 
> class User(Base):
>     __tablename__ = "user"
> 
>     id: Mapped[int] = mapped_column(primary_key=True)
> 
> user = User(id=0)
> assert get_type_hints(user)["id"] == int
> ```
> 
> This came up as I'm actually testing the correspondence of python and sql types, for which I'm using `get_type_hints`.

The change here is because for mixins, we need to maintain the `Mapped[]` aspect of the annotation each time the class is used in a new mapping, so it restores the previous `__annotations__` after applying dataclasses.    So I applied this to the mapped class as well, seeing this as consistent. 

Why would you want it to be `int` when it stated that it was `Mapped[int]`?   I think it's more correct that `__annotations__` illustrates the actual Python annotations that were present.  We can create public API that will give you the other version of it, I would just add them to the class like `__dataclass_annotations__` or something like that.

**Reply by BayerSe** [reply]
> Why would you want it to be `int` when it stated that it was `Mapped[int]`? I think it's more correct that `__annotations__` illustrates the actual Python annotations that were present. We can create public API that will give you the other version of it, I would just add them to the class like `__dataclass_annotations__` or something like that.

This is a test from the "old way", where I wanted to avoid dumb mistakes. It's not strictly necessary anymore today, but I realized that it passed in 2.0.3 but fails in 2.0.4dev0. If I want to keep this test, I can simply move to  `user.__dataclass_fields__["id"].type`, but I wanted to raise the point nevertheless so that you can make an informed decision on that.

**Reply by zzzeek** [reply]
OK I guess I can add a field for this but I would have to decide the best way to namespace it, or make it private with a special function to access it.

**Reply by BayerSe** [reply]
Thanks, but I don't think it's necessary for my use case. I only wanted to highlight a change in behavior between the two versions.

**Reply by CaselIT** [reply]
> what change? thread is too long.

the change you where replying in that comment (Mapped[int] vs int in `__annotations__`)

**Reply by CaselIT** [reply]
>  I would just add them to the class like `__dataclass_annotations__` or something like that.

let's prefix it with `sa` to avoid possible conflicts with 3rd party tools. So something like ` __sa_dataclass_annotations__`

**Reply by CaselIT** [reply]
> Thanks, but I don't think it's necessary for my use case. I only wanted to highlight a change in behavior between the two versions.

I agree, since this is used only for dataclasses and you can use `user.__dataclass_fields__["id"].type` like you mentioned. This is probably the best solution, since we avoid duplicating standard dataclass features

**Reply by zzzeek** [reply]
oh, you can do that?  then yes , we should not change anything then.

**Reply by CaselIT** [reply]
yes, `.type` is documented. `__dataclass_fields__` is not but there is `dataclasses.fields(cls)` that can be used, so should be ok for this use case

**Reply by BayerSe** [reply]
Hi,

I'm sorry to bring this topic up again, but unfortunately, I was able to find a case where `.type` does not reveal the actual type of the dataclass object.

```python
from typing import Annotated
from sqlalchemy.orm import MappedAsDataclass, DeclarativeBase, mapped_column, Mapped
from dataclasses import fields

int_pk = Annotated[int, mapped_column(primary_key=True)]


class Base(MappedAsDataclass, DeclarativeBase):
    ...


class Data(Base):
    __tablename__ = 'data'
    a: Mapped[int_pk]
    b: Mapped[int] = mapped_column(primary_key=True)
```

The reported dataclass types are

```python
> fields(Data)[0].type
typing.Annotated[int, <sqlalchemy.orm.properties.MappedColumn object at 0x7fd0d0a8e0c0>]
``` 

and 

```python
> fields(Data)[1].type
<class 'int'>
```

which is a bit cumbersome / strange, as:

```python
> fields(Data)[0].type is int
False
> fields(Data)[1].type is int
True
```

whereas both types are clearly `int`.

**Reply by CaselIT** [reply]
I think this is not something that should be solved by sqlalchemy, since this is a python typing functionality, and sqlalchemy should not get into typing inspection, at least in the public api


Have tou tried using https://docs.python.org/3/library/typing.html?highlight=annotated#typing.get_type_hints ?

**Reply by BayerSe** [reply]
Thanks, @CaselIT. I think I can work with `get_type_hints`. The output is:
```python
> get_type_hints(Data)
{'a': sqlalchemy.orm.base.Mapped[int], 'b': sqlalchemy.orm.base.Mapped[int]}
```


From that, it seems as if I can access the original type via  `typing.get_args`

```python
> get_args(get_type_hints(Data)["a"])
(<class 'int'>,)
> get_args(get_type_hints(Data)["b"])
(<class 'int'>,)
```

**Reply by CaselIT** [reply]
An alternative is to use `get_args` on `fields(Data)[0].type`, like you would when using `Annotated` in plain dataclasses.
